tools/c7n_azure/c7n_azure/filters.py

```diff
@@ -246,6 +246,35 @@ def __call__(self, i):
             action_date - timedelta(days=self.skew, hours=self.skew_hours))
 
 
+class DiagnosticSettingsFilter(ValueFilter):
+
+    schema = type_schema('diagnostic-settings', rinherit=ValueFilter.schema)
+
+    def validate(self):
+        if not self.manager.resource_type.diagnostic_settings_enabled:
+            raise FilterValidationError("Resource type %s does not have diagnostic settings." % (
+                                        self.manager.type))
+
+    def process(self, resources, event=None):
+        # Process each resource in a separate thread, returning all that pass filter
+        with self.executor_factory(max_workers=3) as w:
```

Comment:
wrt to threading, we've been trying to move towards some consistency around patterns and usage, ie. using concurrent.future submit on a worker pool with a resource set. concurrent futures does imply some overhead, so in general we've been trying to batch/chunk the amount of work for execution. ie in pseudo python code.

```
from c7n.utils chunks

futures = []
results = []

# `with` thread pool context manager - following is indented a block

for resource_set in chunks(resources, 20):
    futures.append(w.submit(self.process_resource_set, client, resource_set))
for f in as_completed(futures):
    if f.exception():
       log()
    else:
       results.extend(f.result())
return results
    
```


one question i had is whats the thread safety around azure client instances, i'm assuming they are thread safe swagger static gen/urllib3/requests for comms.

Reply:
I implemented the threading pattern, but I'll have to get back to you about the thread safety! I'm not sure the answer!